Update standard operations section

manish-pandey-arm opened the pull request:
Provide details around security considerations and add new standard operations like "Removing a TE" and "Overwriting a TE" as mentioned in https://github.com/FirmwareHandoff/firmware_handoff/issues/62

manish-pandey-arm committed:
Add security considerations for Relocating a TL

Signed-off-by: Manish Pandey <manish.pandey2@arm.com>

manish-pandey-arm committed:
Describe "Removing a TE" operation

Signed-off-by: Manish Pandey <manish.pandey2@arm.com>

manish-pandey-arm committed:
Describe "Overwriting a TE" operation

Overwrite operation is used to replace content of a given TE

Signed-off-by: Manish Pandey <manish.pandey2@arm.com>

manish-pandey-arm committed:
Describe helper utility for adding a Void TE

Steps for adding a void TE was already described in creating a new TE
but this operation can be used at multiple places so introduce it as
stand alone operation so that it can be referenced elsewhere.

Signed-off-by: Manish Pandey <manish.pandey2@arm.com>

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+
+#. Mark the TE as void by setting its tag_id to XFERLIST_VOID
+
+#. Erase the TE data if `scrub_data` is TRUE
```
```
*(optional)* Set the `te.data_size` bytes starting at `te_base_addr + te.hdr_size` to 0x0.
```

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+#. Validate the TL header as mentioned in `Validating a TL header`_
+
+#. Locate the `te_base_addr` of the TE to be removed by iterating through the TL.
+
```
To keep the checksum instructions in line with other operations, add here (before making changes):
```
If `has_checksum`, xor the `te.hdr_size + te.data_size` bytes starting at `te_base_addr` with `tl.checksum`
```

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+
+#. Erase the TE data if `scrub_data` is TRUE
+
+#. If checksum is enabled, recalculate and update `tl.checksum`
```
```
If `has_checksum`, xor the `te.hdr_size + te.data_size` bytes starting at `te_base_addr` with `tl.checksum`
```

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+   `old_void_data_size - new_data_size` is greater or equal to `0x8` then call
+   `Adding a void TE`_ with following arguments:
+
+   #. `void.te.base_addr` = `te_base_addr + align8(new_data_size + 0x8)`
```
nit: `void_te`, not `void.te`

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+
+   #. `void.te.base_addr` = `te_base_addr + align8(new_data_size + 0x8)`
+
+   #. `void.te.size` =  `old_void_data_size - align8(new_data_size)`
```
This is missing another `- 0x8`

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -343,6 +343,23 @@ Inputs:
    #. Add `align8(te.hdr_size + te.data_size)` to `te_base_addr`.
 
+
+Adding a void TE
```
I would call this "Overwriting part of the TL with a void TE", because this isn't really "adding" a TE (to the end) the way the normal "Adding a new TE" operation does.

I would maybe put this at the end or at least further down in this section, because it's not really a useful operation on its own, it's only a common step for two others, and it's weird for the reader to see this first.

danh-arm replied:
I kind of agree with this but would go one step further. If we're adding an "Overwriting a TE" section then I think this it's unnecessary to have this operation also embedded within the existing "Adding a TE" section. So I would remove that bit from "Adding a TE", rename it to "Append a TE", and inline this "Overwriting part of the TL with a void TE" operation within the "Overwriting a TE" section.

danh-arm replied:
On 2nd thoughts, you might still need this operation in 2 places ("Overwriting a TE" and "Removing a TE"), in which case it might better in its own section. But still it should be below the operations that use it.

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -343,6 +343,23 @@ Inputs:
+
+Inputs:
+
+- `te_base_addr`: Base address where void TE to be added
```
It might be useful to put the alignment part in here so that it doesn't need to be mentioned in the two calling operations (e.g. call this `start_address`, and then start with calculating `te_base_addr` as `align8(start_address)`)?

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -343,6 +343,23 @@ Inputs:
+
+- `te_base_addr`: Base address where void TE to be added
+- `size`: Total size available for void TE
+
```
You also need to xor the checksum here, to first remove the old state of those bytes from the checksum before adding the new one.

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -343,6 +343,23 @@ Inputs:
+#. Set `te.hdr_size` (`te_base_addr + 0x3`) to `0x8`
+
+#. Set `te.data_size` (`te_base_addr + 0x4`) to `align8(size) - 0x8`
+
```
If you want to mention optional data erasure in the "Remove" operation, you should probably do the same here?  (Note that this will change the area over which the checksum needs to be calculated. In fact, it may be easier to rewrite the "Remove" operation by referencing this one?)

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+
+#. Validate the TL header as mentioned in :ref:`Validating a TL header`
+
+#. Locate the `te_base_addr` of the TE to be overwritten by iterating through the TL
```
Same as above I think this should be the input, and it's odd to mention the validation step here.

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+
+#. Locate the `te_base_addr` of the TE to be overwritten by iterating through the TL
+
+#. Ensure that `te.data_size` is greater or equal to `new_data_size`
```
Rather than just write this I think we should give guidance on how to handle the case when it isn't. I think you could write:
```
#. If `te.data_size` is smaller than `new_data_size`, abort this operation here and instead first follow `Adding a void TE` with `te_base_addr` and `te.data_size`, then follow `Adding a new TE` with `tl_base_addr`, `tag_id`, `new_data_size` and `[data]`.
```

jwerner-chromium replied:
After this step you need to first subtract the old header and data bytes from the checksum if `has_checksum`.

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
 
-   #. Use `te_base_addr + align8(new_data_size + 0x8)` as the new `te_base_addr`
-      for a new XFERLIST_VOID tag.
+#. Erase the old TE data from `te_base_addr + 0x8` until  `te.data_size`
```
This is inefficient. If erasure is (optionally) desired, we only need to overwrite the rest of the old data that remains after new data has been written. I think this should be intergrated into the "Add a void TE" operation, because it is the same requirement in all its callers.

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
 
-   #. If `has_checksum`, xor the 8 bytes from `te_base_addr` to
-      `te_base_addr + 0x8` with `tl.checksum`.
+#. Copy the TE data into `te_base_addr + 0x8`
```
nit: "Copy or generate" to match the wording in "Adding a new TE"

jwerner-chromium commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
 
-   #. If `has_checksum`, xor the 8 bytes from `te_base_addr` to
-      `te_base_addr + 0x8` with `tl.checksum`.
+#. If checksum is enabled, recalculate and update `tl.checksum`
```
This should again explicitly describe how to account for the new bytes after making the change.

manish-pandey-arm commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+Inputs:
+
+- `tl_base_addr`: Base address of the TL from which TE to be deleted
+- `tag_id`: ID number of the tag to be deleted
```
It is implicitly used in finding "te_base_addr", rephrased the sentence.

manish-pandey-arm replied:
used te_base_addr as input

manish-pandey-arm commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+- `tag_id`: ID number of the tag to be deleted
+- `scrub_data`: Boolean option to scrub TE data
+
+#. Validate the TL header as mentioned in `Validating a TL header`_
```
Done

manish-pandey-arm commented on source/transfer_list.rst:
```diff
@@ -391,23 +408,60 @@ Inputs:
+
+#. Validate the TL header as mentioned in `Validating a TL header`_
+
+#. Locate the `te_base_addr` of the TE to be removed by iterating through the TL.
```
Done

manish-pandey-arm commented on source/transfer_list.rst:
```diff
@@ -391,23 +391,52 @@ Inputs:
+   follow `Adding a void TE`_ with `te_base_addr` and `te.data_size`, then follow `Adding a new TE`_
+   with `tl_base_addr`, `tag_id`, `new_data_size` and `[data]`
+
+#. If `has_checksum`, xor the `te.data_size` bytes starting at `te_base_addr + te.hdr_size` with `tl.checksum`
```
My understanding is, An overwrite operation updates the data linked to a tag ID without changing the tag ID itself.

manish-pandey-arm replied:
Thanks for spotting, done!

manish-pandey-arm commented on source/transfer_list.rst:
```diff
@@ -391,23 +391,52 @@ Inputs:
+#. Copy `new_data_size` byte of data starting at `te_base_addr + te.hdr_size`
 
-   #. Set `te.hdr_size` (`te_base_addr + 0x3`) to `0x8`.
+#. If `has_checksum`, xor the `new_data_size` bytes starting at `te_base_addr + te.hdr_size` with `tl.checksum`
```
The API descriptions currently use 0x8 for existing TEs (e.g., VOID_TE), based on their current header size. However, since this may not hold true for all TEs in the future, the generic descriptions refer to te.hdr_size to avoid assumptions. If preferred, I can submit a separate PR to replace all uses with the fixed 0x8 to ensure consistency.

sjg20 commented on source/transfer_list.rst:
```diff
@@ -398,6 +398,19 @@ Inputs:
 
    #. `void_te.size` =  `old_void_data_size - align8(new_data_size + 0x8)`
 
+Removing a TE
```
Seems to be missing a commit message

manish-pandey-arm replied:
Have only got commit header but no body, hope that is fine.

sjg20 replied:
Please add a body to all commits.

jmarinho commented on source/transfer_list.rst:
```diff
@@ -400,6 +400,19 @@ Inputs:
+
+   #. `void_te.base_addr` = `te_base_addr`
+
+   #. `void_te.size` = `te.data_size + te.hdr_size - 0x8`
```
is the second arg void_te.data_size ?

see https://github.com/FirmwareHandoff/firmware_handoff/pull/66/commits/5f526c05259352757d7d7a8c1b4fc7db97e40820#diff-75d1d8518f52351181570a42b6cee154a72b2eb15b3771b969f182b2e537b41eR513

jmarinho commented on source/transfer_list.rst:
```diff
@@ -413,6 +413,35 @@ Inputs:
+#. If `te.data_size - new_data_size` is greater than or equal to `0x8`, then create a new
+   void TE to fill the remaining space by calling `Adding a void TE`_ with following arguments:
+
+   #. `void.te.base_addr` = `te_base_addr + te.hdr_size + align8(new_data_size)`
```
nit: void_te ?

jmarinho commented on source/transfer_list.rst:
```diff
@@ -413,6 +413,35 @@ Inputs:
+
+Inputs:
+
+- `tl_base_addr`: Base address of the TL from which TE to be overwritten
```
nit: I'd argue tl_base_addr is not an argument to the Overwrite TE operation.
The tl_base_addr is used in the "Adding a new TE" procedure, which is executed if the "Overwriting a TE" aborts.

manish-pandey-arm replied:
Yes, you are right. When the overwrite operation returns an abort, the caller may choose to fall back to the "Adding a new TE" procedure, at which point tl_base_addr will be explicitly provided as an argument to that call.

jmarinho reviewed (dismissed):
small set of nits, but otherwise LGTM

harrisonmutai-arm commented on source/transfer_list.rst:
```diff
@@ -507,6 +539,33 @@ Inputs:
+- `te_base_addr`: Base address where void TE to be added
+- `data_size`: Size in bytes of the data to be encapsulated in void TE
+
+#. If `has_checksum`, subtract `te.hdr_size + data_size` bytes starting at `te_base_addr` from `tl.checksum`
```
Should this say "subtract the sum of the bytes starting at `te_base_addr` up to `te_base_addr + te.hdr_size + data_size` from `tl.checksum` instead? Otherwise, this reads as if were subtracting the sum of the header size and data size.

harrisonmutai-arm commented on source/transfer_list.rst:
```diff
@@ -507,6 +539,33 @@ Inputs:
+
+#. If `has_checksum`, subtract `te.hdr_size + data_size` bytes starting at `te_base_addr` from `tl.checksum`
+
+#. Set `te.tag_id` (`te_base_addr + 0x0`) to `0x0` (XFERLIST_VOID)
```
I'm in two minds about providing the offsets explicitly (e.g. `te_base_addr + 0x3`). On one hand it's a useful reference for the implementer, but it may prove a bit of a burden to maintain down the line. As I understand it, the header layout is liable to change. If it does, we'd have to fix all the places in the spec where we reference hard-coded offsets. Given the structure is well defined, I think it's probably sufficient to just refer to it directly (i.e. `te.tag_id`).

manish-pandey-arm replied:
We can do this universally when migrating to appendinx ?

harrisonmutai-arm commented on source/transfer_list.rst:
```diff
@@ -392,23 +392,55 @@ Inputs:
+
+Inputs:
+
+- `te_base_addr`: Base address of the TE to be removed
```
For libtl we use the `tl_base_addr` to perform very naive memory management. If the previous and subsequent TE's are empty we try to coalesce them with the removed TE. This is arguably an optional implementation detail though.

harrisonmutai-arm commented on source/transfer_list.rst:
```diff
@@ -392,23 +392,55 @@ Inputs:
+   follow `Adding a void TE`_ with `te_base_addr` and `te.data_size`, then follow `Adding a new TE`_
+   with `tl_base_addr`, `tag_id`, `new_data_size` and `[data]`
+
+#. If `has_checksum`, subtract the `te.data_size` bytes starting at `te_base_addr + te.hdr_size` from `tl.checksum`
```
Similar comments to below, should this say "sum of `te.data_size` bytes.

harrisonmutai-arm commented on source/transfer_list.rst:
```diff
@@ -392,23 +392,55 @@ Inputs:
 
-   #. Use `te_base_addr + align8(new_data_size + 0x8)` as the new `te_base_addr`
-      for a new XFERLIST_VOID tag.
+   #. `void_te.base_addr` = `te_base_addr + align8(new_data_size + 0x8)`
```
Instead of hard-coding the header size why not just refer to `te.hdr_size`?  It's slightly clearer what the intent is then.

jmarinho reviewed (dismissed):
LGTM

manish-pandey-arm committed:
Replace hardcoded 0x8 with te.hdr_size in standard operations

Updated the "Standard Operations" section to consistently use
`te.hdr_size` instead of the hardcoded `0x8`. Also add a note explaining
the assumption of `te.hdr_size = 0x8` and its potential future
configurability. This improves consistency, avoids magic numbers, and
prepares the format for variable-sized TE headers.

Signed-off-by: Manish Pandey <manish.pandey2@arm.com>

jmarinho reviewed (commented):
LGTM

sjg20 commented on source/transfer_list.rst:
```diff
@@ -417,6 +417,39 @@ Inputs:
 
+Overwriting a TE
+^^^^^^^^^^^^^^^^
+
```
It would be useful to define or motivate this more clearly. Why would one want to do this?

manish-pandey-arm replied:
Overwriting a TE is a space-efficient way to update the data associated with an existing tag without changing its position in the transfer list. This is useful in scenarios where the tag ID remains valid, and only the payload needs to be refreshed. It avoids the overhead of allocating new space and modifying the TL structure unless the new data exceeds the original size, in which case a replacement strategy is triggered.

@sjg20 Do you these comments in spec or you happy with current version ?

manish-pandey-arm replied:
@danh-arm, could you please merge this PR? @sjg20 is happy with the patch, aside from a single minor nit.

sjg20 reviewed (dismissed):
LGTM with one comment

danh-arm commented on source/transfer_list.rst:
```diff
@@ -507,6 +507,9 @@ Inputs:
 
+.. note::
+   After relocating a TL, implementations should consider scrubbing the old TL memory if it contains
+   any secrets that might be accessible to later untrusted software.
```
minor: I think this note also applies in the case of Removing a TE or Overwriting a TE. This could perhaps be duplicated or placed in a more general section?

danh-arm reviewed (approved):
Just 1 minor comment that can be addressed later
